Add OutputReport implementation that generates an HTML report

pocmo opened the pull request:
Fixes #490

I didn't want to include some big HTML library or whatever. So I wrote my own little helper ("htmlSnippet")

A generated report looks like this:

![screenshot-2017-11-30 detekt report](https://user-images.githubusercontent.com/89638/33437408-1f5c4fce-d5e8-11e7-9666-df308e349acf.png)

Not super beautiful. But hopefully a good start. :)

schalkms commented on detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/HtmlOutputReport.kt:
```diff
@@ -0,0 +1,102 @@
+package io.gitlab.arturbosch.detekt.cli.out
+
+import io.gitlab.arturbosch.detekt.api.*
```
Can we please remove the wildcard import.

pocmo replied:
Argh. I hate that IntelliJ does this by default.

schalkms replied:
No problem ;)
Detekt has a `idea-code-style.xml` in its root directory.

schalkms commented on detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/HtmlOutputReport.kt:
```diff
@@ -0,0 +1,102 @@
+	override fun render(detektion: Detektion) =
+			ClasspathResourceConverter().convert(DEFAULT_TEMPLATE).openStream().bufferedReader().use { it.readText() }
+					.replace("@@@metrics@@@", renderMetrics(detektion.metrics))
+					.replace("@@@findings@@@", renderFindings(detektion.findings))
```
I like the template mechanism. Can we refactor this to a constant for readability reasons?

schalkms commented on detekt-cli/src/test/kotlin/io/gitlab/arturbosch/detekt/cli/out/HtmlOutputFormatTest.kt:
```diff
@@ -0,0 +1,57 @@
+				CodeSmell(Issue("id_b", Severity.CodeSmell, "A2"), entity2, message = "B2"),
+				CodeSmell(Issue("id_c", Severity.CodeSmell, "A3"), entity3, message = "B3")))
+
+		assertThat(result).contains("<span class=\"location\">\nsrc/main/com/sample/Sample1.kt:11:1\n</span>")
```
For readability reasons I think we should refactor this into a separate method.

schalkms commented on detekt-cli/src/test/kotlin/io/gitlab/arturbosch/detekt/cli/out/HtmlOutputFormatTest.kt:
```diff
@@ -0,0 +1,57 @@
+
+	@Test
+	fun testRenderMultipleSmells() {
+		val entity1 = Entity("Sample1", "com.sample.Sample1", "",
```
For readability reasons I think we should refactor this into a separate method.

schalkms commented on detekt-cli/src/test/kotlin/io/gitlab/arturbosch/detekt/cli/out/HtmlOutputFormatTest.kt:
```diff
@@ -0,0 +1,57 @@
+						"efgh", "src/main/com/sample/Sample2.kt"))
+		val entity3 = Entity("Sample2", "com.sample.Sample3", "",
+				Location(SourceLocation(33, 3), TextLocation(0, 30),
+						"ijkl", "src/main/com/sample/Sample3.kt"))
```
For readability reasons I think we should refactor this into a separate method.

schalkms commented on detekt-cli/src/test/kotlin/io/gitlab/arturbosch/detekt/cli/out/HtmlOutputFormatTest.kt:
```diff
@@ -0,0 +1,57 @@
+
+		assertThat(result).contains("<span class=\"description\">\nA1\n</span>")
+		assertThat(result).contains("<span class=\"description\">\nA2\n</span>")
+		assertThat(result).contains("<span class=\"description\">\nA3\n</span>")
```
For readability reasons I think we should refactor this into a separate method.

Mauin commented on detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/out/HtmlOutputReport.kt:
```diff
@@ -0,0 +1,102 @@
+		}
+	}
+
+	private fun htmlSnippet(init: HTMLSnippet.() -> Unit): String {
```
Does it make sense extracting these HTML generating methods into their own file?

pocmo replied:
Maybe. Of course I only implemented the bare minimum needed for the current report. :)

Does it make sense to move this separate file into ...detekt.cli.out too or is there a better place?

pocmo commented:
@schalkms @Mauin Thank you for the reviews. I think I addressed everything and updated the PR.

@Mauin I moved the HtmlSnippet code to its own file. With that I added some tests for this helper too. :)

arturbosch commented:
Cool, thanks for the contribution!
